### ISSUE TITLE
Licensing: proposed addition with an Apache 2.0 with LLVM runtime exception 

### ISSUE THREAD

**Issue by rgommers** [scipy/scipy#20396]
Following up on the licensing part of the proposed vendoring of `kokkos/mdspan`, which has an Apache 2.0 + LLVM runtime exception license (see gh-20334, https://github.com/scipy/scipy/pull/20320#pullrequestreview-1969858082).

@steppi opened https://github.com/kokkos/mdspan/issues/321 upstream to inquire about relicensing, which _may_ be possible but isn't a fast decision apparently.

First some context (copied from the review comment linked above):

The Apache 2.0 license in `mdspan.h` seems incompatible with SciPy's BSD-3 license. See:
- http://scipy.github.io/devdocs/dev/core-dev/index.html#licensing
- http://scipy.github.io/devdocs/dev/hacking.html#license-considerations
- https://github.com/scipy/scipy/issues/8370#issuecomment-363592943

It's not completely impossible to reconsider the current ban on Apache 2.0, as indicated by the comment in the last link above, but it's a major decision with long-term consequences that, even if we drop GPLv2 compatibility, will complicate the project's licensing situation.

* * *

After reviewing the LLVM runtime exception clause, it seems clear that that does address the GPLv2 compatibility issue, so there is no problem there.

The other problem is source-level licensing. Right now we have a BSD-3 license for SciPy as a whole (see [LICENSE.text](https://github.com/scipy/scipy/blob/main/LICENSE.txt)), and a large-ish set of licenses for vendored code (see [LICENSES_bundled.txt](https://github.com/scipy/scipy/blob/main/LICENSES_bundled.txt)). All of the licenses for bundled code are BSD-3 compatible: various other BSD flavors, MIT, Boost, and Qhull (BSD-like) licenses.

The LLVM runtime exception part of the `kokkos/mdspan` license is only related to linking. So for the license expression one would need to use to describe all of SciPy, that's not very relevant. Introducing Apache 2.0 would complicate describing the license - but perhaps not by that much. Distros tend to already exhaustively search for licensing and copyright info and mash everything together ([example for Debian](https://metadata.ftp-master.debian.org/changelogs//main/s/scipy/scipy_1.10.1-2_copyright)). So while not ideal to add this new license, it _may_ be acceptable.

Let me ask for some input here. @rkern any thoughts? @sandrotosi as a Debian packager, would you see any issues if we'd accept this extra license?

**Comment by tylerjereddy** [comment]
Yeah, Robert usually has licensing insights I don't think of. One thing I'd say is that another NumFOCUS project I work on recently went through a license change process and it was pretty brutal, like emailing every committer. Hopefully it wouldn't be that intrusive if we needed to change something.

**Comment by sandrotosi** [comment]
i am no longer involved in packaging scientific modules for Debian, so im afraid i cant provide any input.

**Comment by rkern** [comment]
I think @rgommers covered everything I can think of. The LLVM Exceptions cover the GPLv2 incompatibility that has mostly been our hard objection to Apache 2.0 code. I *think* it works, from a layman's reading, but it's entirely possible actual lawyers have opinions on whether or not that precise language actually has any real effect. Given its provenance, I'd wager that it's fine, but that's all I'm basing that on.

**Comment by tupui** [comment]
NumFOCUS can provide this service (it's one the service we are paying for). Might be worth asking them so that it also cover more the project's liability if we do move forward with the proposal.

**Comment by izaid** [comment]
Thanks @rgommers, @rkern, and @tupui for discussing and considering this. I do think it is a worthwhile thing to pursue.

In the short term, I've written my own poor man's implementation of `mdspan` to unblock the gufunc PR. I'll be checking that in very soon. I still think it's worth pursuing this licensing discussion, as the Apache 2.0 (now with LLVM exception) has come up before and I'd really much prefer to use the reference implementation.

**Comment by rgommers** [comment]
Proposed resolution here:
1. Using _Apache 2.0 + LLVM runtime exception_ licensed header-only code in SciPy is fine.
2. Using _Apache 2.0_ licensed code in SciPy is unchanged (still not allowed), but can be addressed/reconsidered once we have a compelling need for it.

**Comment by izaid** [comment]
> The suggestion made in [#20320 (comment)](https://github.com/scipy/scipy/pull/20320#issuecomment-2048673569) was to not use `kokkos/mdspan` but the `CCCL::libcudacxx`. If that's judged a better route, the request to the Kokkos team can be closed.

Just want to mention that the `CCCL::libcudacxx` mdspan _is_ the kokkos/mdspan, see https://github.com/NVIDIA/cccl/blob/2889c3c79c53e4db6394ca86bf09d60ff1fb4897/libcudacxx/include/cuda/std/detail/libcxx/include/__mdspan/mdspan.h#L45

This kokkos/mdspan reference implementation has gone into most compilers and libraries, so far as I can see.

**Comment by lucascolley** [comment]
> For kokkos/mdspan it'd be a relicensing, but that'd be a single entity signing off on it, not a ton of individuals.

> Using Apache 2.0 + LLVM runtime exception licensed header-only code in SciPy is fine.

I don't understand how these two statements gel with the response from kokkos:

> Note that we are under Apache-2 with LLVM Exceptions. Thus no attribution is required in binary code, and you just need to not delete our license notice in headers you copy.

What relicensing is required from them?

**Comment by rgommers** [comment]
> What relicensing is required from them?

From the issue description: "@steppi opened [kokkos/mdspan#321](https://github.com/kokkos/mdspan/issues/321) upstream to inquire about relicensing, which _may_ be possible but isn't a fast decision apparently."

If the answer to that had been yes, there would have been no need for this whole discussion.

**Comment by rgommers** [comment]
> Just want to mention that the `CCCL::libcudacxx` mdspan _is_ the kokkos/mdspan,

Extended/updated for better CUDA support I assume, right? If it's identical, there would be no need for the suggestion. My point was that if it's extended by NVIDIA and then we use that version, then the original `kokkos/mdspan` being relicensed wouldn't have helped us.

**Comment by ilayn** [comment]
By the way, if it is going to be too much of an issue we can get it from the backported C++ version. This is really not a central component for our code. We are really not bound to use any of this if push comes to shove but would be nice if we can.

**Comment by izaid** [comment]
> > Just want to mention that the `CCCL::libcudacxx` mdspan _is_ the kokkos/mdspan,
> 
> Extended/updated for better CUDA support I assume, right? If it's identical, there would be no need for the suggestion. My point was that if it's extended by NVIDIA and then we use that version, then the original `kokkos/mdspan` being relicensed wouldn't have helped us.

Ah, ok, I'm with you now. Yes, indeed.

**Comment by leofang** [comment]
> @leofang confirmed in [#20320 (comment)](https://github.com/scipy/scipy/pull/20320#issuecomment-2048673569) that the situation we're considering here - which is essentially the same for CuPy (Apache 2.0 with LLVM exception + MIT) aside from GPLv2 concern - was reviewed by NVIDIA's legal team and found to be okay.

Just trying to be cautious and clear: the case that our legal checked was CuPy (MIT) vendoring CCCL (Apache 2.0 with LLVM exception). CuPy unlike SciPy does not have any GPLv2 code so the situation was simpler. Any use case beyond that was not covered in the case study and I would advise for seeking legal advices, as I obviously can't speak for our legal 🙂

**Comment by rgommers** [comment]
Sure, I understand, thanks @leofang. That's why I said "essentially the same ... aside from GPLv2 concern". I was more concerned about the other parts (source level redistribution, and "did we miss anything else"). 

> CuPy unlike SciPy does not have any GPLv2 code so the situation was simpler. 

SciPy itself also doesn't have any GPLv2 code anywhere. The constraint is more about the ability of a third party to mix in GPLv2-licensed code and distribute SciPy together with a binary built from that other code. Which is not fine with plain Apache 2.0, but fine with the LLVM exception.

**Comment by rkern** [comment]
As a rule, I consider the actions of large corporations to provide almost no information about what's legally good for us. Legal advice is provided in a context specific to the individual asking (and paying) for that advice. We do not know the actual details of the question given to the NVIDIA legal team, we do not know the details of their response, we do not know what NVIDIA's risk profile is (except that it surely isn't ours), we do not know the goals of NVIDIA. Legal advice is only sometimes a clearcut, universal "this is okay for anyone to do" but usually more "given what you want and ignoring what you don't care about, and given the amount of money you are willing to put at risk in future legal battles, this is okay for _you_ to do."

We already knew that we _could_ mix Apache 2.0 and our BSDish code (i.e. that there isn't a specific incompatibility of terms between the two like there is between plain Apache 2.0 and GPLv2), and that's probably the extent that I'd trust NVIDIA's examination of the matter. I don't think that we can take that as an answer to "did we miss anything else".

**Comment by steppi** [comment]
Has anyone followed up to ask NumFocus for legal advice on this matter? The immediate concern is resolved, because @izaid wrote his own minimal implementation for use in `special`, but it would be nice to have the full implementation available for use in C++ code in SciPy more generally.

**Comment by rgommers** [comment]
> Has anyone followed up to ask NumFocus for legal advice on this matter?

I have not, since I don't see the need. My proposed resolution posted above (https://github.com/scipy/scipy/issues/20396#issuecomment-2049253707) is fine I think.

The few comments on the relevance of what NVIDIA legal thinks side-tracked us a bit, but I think the situation is clear either way. As an extra data point not mentioned before: we have been shipping Apache 2.0-licensed code (no LLVM runtime exception, plain old Apache 2.0) in our sdist and repo for many years (namely in `doc/source/_static/scipy-mathjax`) and we've never had a complaint. So I don't think there's anything new here, the question is resolved for me.

**Comment by steppi** [comment]
Thanks @rgommers and @rkern. I wasn't sure if we'd come to a final resolution. I understand now @rkern, it does make sense to point out that the conclusions from NVIDIA's lawyers aren't necessarily relevant to SciPy, and make that part of the conversational record.

@ev-br, @izaid, it looks like we can start using either Kokkos or NVIDIA's `mdspan` in SciPy. I guess we just need to add an entry for `mdspan` to https://github.com/scipy/scipy/blob/main/LICENSES_bundled.txt.

**Comment by rgommers** [comment]
> But I want it to be clear in the conversation record _why_ it's fine, so if points are being raised to support that position that I don't think accomplish that, I'm going to comment on them.

Yes agreed, this is valuable. We can simply discard the NVIDIA-legal part of this conversation as not relevant.

To recap the reasons:

1. For binary artifacts: after reviewing the LLVM runtime exception clause, it seems clear that that does address the GPLv2 compatibility issue, so there is no problem there. 
2. For source-level licensing: after review, there does not seem to be an issue here. We are shipping Apache licensed code in the sdist already, an entry in `LICENSES_bundled.txt` should suffice. No one has suggested a concrete reason for why this isn't fine. The initial concern was mine, and I'm happy after considering it more closely.

> I do agree that it's fine to start including `Apache-2.0 WITH LLVM-exception` code.

Great. I suggest we consider this resolved then. Final outcome:

1. Using _Apache 2.0 + LLVM runtime exception_ licensed header-only code in SciPy is fine.
2. Using _Apache 2.0_ licensed code in SciPy is unchanged (still not allowed), but can be addressed/reconsidered once we have a compelling need for it.
